tests/workflows/test_pytorch_optuna.py

```diff
@@ -0,0 +1,208 @@
+        )
+        device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+        # Create models
+        generator = get_generator(trial).to(device)
+        discriminator = get_discriminator(trial).to(device)
+
+        # Establish convention for real and fake labels during training
+        real_label = 1.0
+        fake_label = 0.0
+
+        # Setup loss criterion / optimizers and train
+        loss = nn.BCELoss()
+        optimizer_discriminator = optim.Adam(
+            discriminator.parameters(), lr=0.00001, betas=(0.5, 0.999)
+        )
+        optimizer_generator = optim.Adam(
+            generator.parameters(), lr=0.00001, betas=(0.5, 0.999)
+        )
+        num_epochs = 5
+        for epoch in range(num_epochs):
+            # For each batch in the dataloader
+            for batch in dataloader:
+                # Train discriminator with all-real batch (maximize log(D(x)) + log(1 - D(G(z))))
+                discriminator.zero_grad()
+                # Format data batch
+                real_cpu = batch[0].to(device)
+                batch_size = real_cpu.size(0)
+                label = torch.full([batch_size], real_label, device=device)
+                # Forward pass real batch through D
+                output = discriminator(real_cpu).view(-1)
+                loss_real = loss(output, label)
+                loss_real.backward()
+
+                # Train with all-fake batch
+                # Generate fake image batch with G
+                noise = torch.randn(batch_size, NZ, 1, 1, device=device)
+                fake = generator(noise)
+                label.fill_(fake_label)
+
+                # Classify all fake batch with D
+                output = discriminator(fake.detach()).view(-1)
+                loss_fake = loss(output, label)
+                loss_fake.backward()
+
+                # Compute error of discriminator as sum over the fake and the real batches
+                loss_discriminator = loss_real + loss_fake
+                optimizer_discriminator.step()
+
+                # Update generator: maximize log(D(G(z)))
+                generator.zero_grad()
+                # fake labels are real for generator cost
+                label.fill_(real_label)
+                # Since we just updated D, perform another forward pass of all-fake batch through D
+                output = discriminator(fake).view(-1)
+                loss_generator = loss(output, label)
+                loss_generator.backward()
+
+                # Update G
+                optimizer_generator.step()
+
+            # Handle pruning based on the intermediate value
+            score = loss_discriminator.item()
+            trial.report(score, epoch)
+            if trial.should_prune():
+                raise optuna.exceptions.TrialPruned()
+
+        return score
+
+    study = optuna.create_study(
+        storage=optuna.integration.DaskStorage(), direction="minimize"
+    )
+
+    n_trials = 100
```

Comment:
This number of trials is much larger than the possible combinations of `trial.suggest_*` options in the objective function. Are there other hyperparameters that are commonly tuned?

Reply:
https://github.com/coiled/benchmarks/pull/787/commits/55d8c1906d27c110734c637dd5a4594a789c79e5